src/sage/schemes/elliptic_curves/ell_number_field.py

```diff
@@ -4062,3 +4068,264 @@ def rational_points(self, **kwds):
+        + 3410611518671/161051*x^20 - 6947130862854/161051*x^19 + 10098387132407/161051*x^18 - 8011521604618/161051*x^17
+        - 509395128662/14641*x^16 + 32582181402218/161051*x^15 - 62691210143713/161051*x^14 + 73259301822126/161051*x^13
+        - 46379791168053/161051*x^12 - 6277600417172/161051*x^11 + 46742832789352/161051*x^10 - 49641085864608/161051*x^9
+        + 2343504014803/14641*x^8 - 861578737314/161051*x^7 - 12167703919007/161051*x^6 + 14278547318070/161051*x^5
+        - 9423834664881/161051*x^4 + 3112085963896/161051*x^3 - 457632005824/161051*x^2 + 30845635072/161051*x + 122023936/161051
+        sage: _modular_polynomial('nonsplit', 7, j)
+        400320064/161051*x^21 + 13029623152/161051*x^20 + 228088753900/161051*x^19 + 2408913460821/161051*x^18
+        + 16642817648786/161051*x^17 + 80889087315407/161051*x^16 + 291040779803200/161051*x^15 + 801917686416123/161051*x^14
+        + 1729396406547138/161051*x^13 + 2959141014912537/161051*x^12 + 4049663724749832/161051*x^11
+        + 4450611572214360/161051*x^10 + 3931865527421272/161051*x^9 + 2786855392905248/161051*x^8
+        + 1576387623537152/161051*x^7 + 704811439294144/161051*x^6 + 22297037611520/14641*x^5 + 5896572491264/14641*x^4
+        + 12558313746944/161051*x^3 + 152158103552/14641*x^2 + 136821293056/161051*x + 5155295232/161051
+
+    TESTS::
+
+        sage: _modular_polynomial('banana', 7, j)
+        Traceback (most recent call last):
+        ...
+        ValueError: (group, p) must be one of ...
+        sage: _modular_polynomial('borel', 101, j)
+        Traceback (most recent call last):
+        ...
+        ValueError: (group, p) must be one of ...
+    """
+    valid_parameters = [("borel", 3), ("borel", 5), ("borel", 7),
+                       ("split", 3), ("split", 5), ("split", 7),
+                       ("nonsplit", 5), ("nonsplit", 7)]
+    if (group, p) not in valid_parameters:
+        raise ValueError("(group, p) must be one of {}".format(valid_parameters))
+
+    from sage.rings.polynomial.polynomial_ring import polygen
+    x = polygen(j.parent())
+    if group == "borel":
+        from sage.schemes.elliptic_curves.isogeny_small_degree import Fricke_polynomial
+        return Fricke_polynomial(p)(x)-j*x
+    if group == "split":
+        if p == 3:
+            return (x-9)**3*(x+3)**3-j*x**3
+        if p == 5:
+            return ((x**2-5)*(x**2+5*x+10)*(x+5))**3-j*(x**2+5*x+5)**5
+        if p == 7:
+            return ((x**2-5*x+8)*(x**2-5*x+1)*(x**4-5*x**3+8*x**2-7*x+7)*(x+1))**3*x-j*(x**3-4*x**2+3*x+1)**7
+    if group == "nonsplit":
+        if p == 5:
+            return 5**4*(2*x+1)*(x+1)**3*(6*x**2+21*x+19)**3-j*(x**2+x-1)**5
+        if p == 7:
+            return ((4*x**2+5*x+2)*(x**2+3*x+4)*(x**2+10*x+4)*(3*x+1))**3-j*(x**3+x**2-2*x-1)**7
+
+    raise RuntimeError # cannot happen -- the assertion would have failed.
```

Comment:
I find there is some utility in this as it would catch bugs (or at least make it easier to see where they are coming from) by raising a specific type of exception rather that silently returning `None`.

Reply:
but this line cannot be reached at all, I think

Reply:
That's the point. It should never be reached. I am treating it like an `assert` statement; something that should never be false/reached. Actually, would that be okay with you: `assert False`?

Reply:
I am sorry, but I do not see the point. We are first checking the input parameters to belong to an explicit list. Then, for each member of the list, the code path ends by a return. So this *proves* that the final line can *never* happen.

Reply:
You're exactly right that this code is _currently_ bug-free. Yet, suppose I change this method, but I introduce a bug for some case. Then I run this code, and without this `raise`, the code silently returns a wrong answer of `None`. Now this is highly likely to fail, but somewhere else. However, the failure is not guaranteed (no reasonable person should trap a `RuntimeError` or an `AssertionError`), which means a wrong answer could actually be returned. Thus, this one extra line makes the code more robust against future errors that could be introduced.

Reply:
ok, ok. I do not agree, but will not oppose any longer, for doc and code inside a hidden method. Good for me